Enable autograd to recognize the XLA backend as one providing multiple devices

dlibenzi opened the pull request:
…e devices, while not being CUDA/HIP.

apaszke reviewed (changes requested):
All places where we do `x.is_cuda() || x.is_xla()` should probably be changed to `!x.is_cpu()`, or even better we should refactor autograd to stop using `int` to signify devices, and use `at::Device` instead.

dlibenzi commented:
> All places where we do `x.is_cuda() || x.is_xla()` should probably be changed to `!x.is_cpu()`, or even better we should refactor autograd to stop using `int` to signify devices, and use `at::Device` instead.

Agreed. !CPU is easy. Let me see what it means to go straight for Device use ...

dlibenzi commented:
> > All places where we do `x.is_cuda() || x.is_xla()` should probably be changed to `!x.is_cpu()`, or even better we should refactor autograd to stop using `int` to signify devices, and use `at::Device` instead.
> 
> Agreed. !CPU is easy. Let me see what it means to go straight for Device use ...

The change to !cpu was easy and made the PR simpler.
Using at::Device is quite more involved as the surface is much bigger.

dlibenzi commented:
> > > All places where we do `x.is_cuda() || x.is_xla()` should probably be changed to `!x.is_cpu()`, or even better we should refactor autograd to stop using `int` to signify devices, and use `at::Device` instead.
> > 
> > 
> > Agreed. !CPU is easy. Let me see what it means to go straight for Device use ...
> 
> The change to !cpu was easy and made the PR simpler.
> Using at::Device is quite more involved as the surface is much bigger.

Except that there is no is_cpu() ....

apaszke commented:
@dlibenzi it's because `is_cpu()` used to be synonymous to `!is_cuda()` 😄 I don't think there's any harm in adding it.

dlibenzi commented:
> > > > All places where we do `x.is_cuda() || x.is_xla()` should probably be changed to `!x.is_cpu()`, or even better we should refactor autograd to stop using `int` to signify devices, and use `at::Device` instead.
> > > 
> > > 
> > > Agreed. !CPU is easy. Let me see what it means to go straight for Device use ...
> > 
> > 
> > The change to !cpu was easy and made the PR simpler.
> > Using at::Device is quite more involved as the surface is much bigger.
> 
> Except that there is no is_cpu() ....

Changed: tensor.is_cpu() -> tensor.device().is_cpu()

dlibenzi commented:
> > > > > All places where we do `x.is_cuda() || x.is_xla()` should probably be changed to `!x.is_cpu()`, or even better we should refactor autograd to stop using `int` to signify devices, and use `at::Device` instead.
> > > > 
> > > > 
> > > > Agreed. !CPU is easy. Let me see what it means to go straight for Device use ...
> > > 
> > > 
> > > The change to !cpu was easy and made the PR simpler.
> > > Using at::Device is quite more involved as the surface is much bigger.
> > 
> > 
> > Except that there is no is_cpu() ....
> 
> Changed: tensor.is_cpu() -> tensor.device().is_cpu()

Broke MSNPU:
```
Mar 10 16:06:43 ======================================================================
Mar 10 16:06:43 ERROR: test_backwards (__main__.TestMSNPUTensor)
Mar 10 16:06:43 ----------------------------------------------------------------------
Mar 10 16:06:43 Traceback (most recent call last):
Mar 10 16:06:43   File "test_cpp_extensions.py", line 673, in test_backwards
Mar 10 16:06:43     d.backward()
Mar 10 16:06:43   File "/opt/python/nightly/lib/python3.7/site-packages/torch/tensor.py", line 106, in backward
Mar 10 16:06:43     torch.autograd.backward(self, gradient, retain_graph, create_graph)
Mar 10 16:06:43   File "/opt/python/nightly/lib/python3.7/site-packages/torch/autograd/__init__.py", line 93, in backward
Mar 10 16:06:43     allow_unreachable=True)  # allow_unreachable flag
Mar 10 16:06:43 RuntimeError: vector::_M_range_check: __n (which is 2) >= this->size() (which is 1)
Mar 10 16:06:43 
Mar 10 16:06:43 ----------------------------------------------------------------------
```

At this point I am going back to is_xla().

dlibenzi commented:
I added an API on the autograd side, to fetch the device, which switches on backend.

apaszke reviewed (changes requested):
Actually now that I look at it, why do we need this change? You know that we'll only spawn as many threads as there are CUDA devices, so in particular if someone has no CUDA-capable GPUs using TPU device indices will cause out of bounds indexing of work queues

dlibenzi commented:
> Actually now that I look at it, why do we need this change? You know that we'll only spawn as many threads as there are CUDA devices, so in particular if someone has no CUDA-capable GPUs using TPU device indices will cause out of bounds indexing of work queues

With that change we spawn as many threads as GPU or TPU (if available) devices.
Without the change:

1. Autograd breaks with multiple devices as the device guard is not properly initialized

2. Not enough threads to process the requests

dlibenzi commented:
Tried to rebase but no luck.
We still have these weird cmake invalid character error when trying to parse the environment.
This is happening randomly on other PRs as well:

https://ci.pytorch.org/jenkins/job/caffe2-builds/job/py2-clang7-rocmdeb-ubuntu16.04-build/19628/console

ailzhang commented:
@pytorchbot rebase this please

ailzhang commented:
The test_dataloader errors should be gone after rebase. cc: @bddppq for rocm build :)

ezyang commented:
I have no idea why ROCm is failing. Maybe you are getting unlucky on a machine.

However, the lint failure is real, and should be fixed.

ezyang commented:
@pytorchbot rebase this please

ezyang commented:
Oh, you got trolled. The non-ASCII character is because your commit message has a Unicode ellipses in it. If you force push it away the problem will go away. I filed #17880 to track this.

dlibenzi commented:
> Oh, you got trolled. The non-ASCII character is because your commit message has a Unicode ellipses in it. If you force push it away the problem will go away. I filed #17880 to track this.

Good catch! Thanks!
Oh, not the commit message .. the github one. The commit message does not have ellipses.
Let me fix it...

dlibenzi commented:
I fixed the CLANG issue as well, even though my change did not affect it before.
Simple change. The set_data() now takes a const& instead of value, to please clang-tidy.

ailzhang commented:
@apaszke  @gchanan  would you mind taking another look at this PR?

gchanan commented on aten/src/ATen/Context.h:
```diff
@@ -76,6 +77,9 @@ class CAFFE2_API Context {
     return detail::getHIPHooks().hasHIP();
   }
+  bool hasXLA() const {
+    return detail::getXLAHooks() != nullptr;
```
Why doesn't this match the other hooks, which are instantiated with versions of the hooks that can return false?

dlibenzi replied:
HIP/CUDA works different from XLA, as far as "bring up" POV.
PT loads CUDA/HIP when it boots. The notion of has-HIP, has-CUDA is embedded in the build.
We load PT, then install our hooks.
The very same version of PT can load w/out XLA.
We could install a dummy version of the XLA hooks, even though there is no XLA, if that's what you prefer.

gchanan replied:
ya, that's what I prefer.

Also, PT doesn't load CUDA/HIP when it boots, only when code requiring it is exercised.

dlibenzi replied:
Applied hooks style similar to CUDA/HIP.

gchanan commented on aten/src/ATen/Context.h:
```diff
@@ -204,7 +208,15 @@ static inline bool hasHIP() {
 static inline size_t getNumGPUs() {
+  auto xla_hooks = detail::getXLAHooks();
+  if (xla_hooks != nullptr) {
+    return xla_hooks->num_devices();
```
this means essentially I cannot have both XLA devices and GPU devices in the same pytorch instance (or at least, I won't have enough autograd threads).  We make this assumption for HIP because that's a limitation of how that specific part of PyTorch is written, but I'm not sure we really want to encode that elsewhere.

Is there anything preventing XLA and GPUs from coexisting?  Can we just add a new function getTotalNonCPUDevices or something that autograd can call?

dlibenzi replied:
Good point, but unfortunately it gets tricky WRT autograd, and its device->queue indexing if we start support GPU and XLA.
It's even more complex since XLA itself, can lower for GPU :)

gchanan replied:
well we'd need to index with actual devices, right?

> It's even more complex since XLA itself, can lower for GPU :) 

what does this mean exactly?

dlibenzi replied:
Look at how the thread initializes.
Each thread sets up a device guard with specific device type.
That guard is then used in other places.
If we allow multiple devices PT GPU together with PT XLA, the autograd code will have to go under a pretty major change, as autograd lower lever code seems only to deal with device indices (and assume CUDA).

XLA internally can use GPU in theory, via the XLA paths (not PT).
From a user POV you could be using GPU, by using XLA, w/out noticing (same interface).
XLA also handle CPU, this is the way we do most of the development in our workstations.

gchanan replied:
> If we allow multiple devices PT GPU together with PT XLA, the autograd code will have to go under a pretty major change, as autograd lower lever code seems only to deal with device indices (and assume CUDA).

Sure, that makes sense.  I can try to see how feasible making that kind of change is.

> XLA internally can use GPU in theory, via the XLA paths (not PT).
From a user POV you could be using GPU, by using XLA, w/out noticing (same interface).
XLA also handle CPU, this is the way we do most of the development in our workstations.

Thanks, that makes sense.  I guess the question is what is the right thing to do from autograd's perspective.  Is the assumption in this world that XLA controls all the (available) devices, so it would be bad perf wise or even invalid if the user is also directly using the GPUs via PyTorch?

dlibenzi replied:
The right, long term, change for autograd would be to make it reason about Device everywhere. Like Adam was mentioning.
I took a quick peek at it, and mong the places where device guard info are used, and the places where only the device ordinal travels, seems pretty big change to me.

gchanan commented on aten/src/ATen/detail/XLAHooksInterface.cpp:
```diff
@@ -0,0 +1,22 @@
+
+}  // namespace
+
+const XLAHooksInterface* getXLAHooks() {
```
as above, can we follow the HIP pattern?

dlibenzi replied:
Tried to explain above, so maybe I can close this one and keep single thread.

gchanan commented on torch/csrc/autograd/input_metadata.h:
```diff
@@ -6,6 +6,14 @@
 namespace torch { namespace autograd {
 
+inline int get_tensor_device(const at::Tensor& tensor) {
+  if (tensor.is_cuda() ||
```
would it  be difficult to just rewrite the places where you need this to do this all in terms of Devices (`tensor.device()`) ?

dlibenzi replied:
This knowledge of "has device", and its way to check, is typical of autograd, so I think it makes sense to have a gating API on the autograd side.

gchanan commented on torch/csrc/autograd/python_hook.cpp:
```diff
@@ -169,15 +169,12 @@ static void check_single_result(PyObject* _original, PyObject* _result, PyObject
   }
 
-  if (original.is_cuda() != result.is_cuda()) {
+  if (original.type().backend() != result.type().backend()) {
```
I don't really understand what this test is doing, because I think the `type().ID()` test above would catch this.  But in any case, if you want the equivalent behavior this should be written in terms of devices, not backends.

dlibenzi replied:
I have tried to minimize the changes in autograd, since I am not too familiar with it.
I had to change it since otherwise it'd fail when used with XLA devices/backends.

gchanan replied:
is there something I can run to see the failure?  My take is the existing code is bogus and thus your code shouldn't matter, but I must be misunderstanding something if you can trigger it.

dlibenzi replied:
Let me try again on this side, with the new code. Might be stray stuff, before I set up correctly the device guard.

dlibenzi replied:
Did not trigger with proper device guard setup.
Removed this change with the new commit.

dlibenzi committed:
Enable autograd to recognize the XLA backend as one providing multiple devices.

gchanan commented:
FYI @ezyang says he's implemented the correct autograd changes before to handle different devices, but they didn't get merged, so he's going to take a stab at that part.

ezyang commented:
I'm going to fix this diff to do it properly. Please hold.

ezyang committed:
Revert Context querying XLA devices.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Rewrite the autograd strategy for XLA.

- Now we use DeviceGuardImpl as the hooks mechanism; no need
  for an extra hooks object (and it wasn't the right format
  anyway)

- Compute the high watermark of devices over all device types,
  and colocate GPUs with the same device index on the same
  autograd thread.  It doesn't have to be this way, but this
  minimized the changes.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
I've pushed the updated patch. It replaces your hooks interface with a standard hooks interface which I think you will find generally useful. In your out-of-tree XLA code, you need to implement the interface `DeviceGuardImplInterface` and register it as `C10_REGISTER_GUARD_IMPL(XLA, XLAGuardImpl);`. See `c10/cuda/impl/CUDAGuardImpl.h` and its cpp file for an example of how to do this.

ezyang committed:
Fix use of non-existent property

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

apaszke commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -352,7 +379,7 @@ static void validate_outputs(const edge_list& edges, variable_list& grads, const
       AT_ERROR(format_error(ss.str()));
     }
-    const auto output_device = output.is_cuda() ? output.get_device() : -1;
+    const int output_device = assign_tensor_to_autograd_thread(output);
```
Well that seems weird. If you have a function called `assign_tensor_to_autograd_thread`, then it feels weird to use it to validate if the output is on a correct device. Someone will change this function in the future, and will not expect that they should look into `validate_outputs` as well...

apaszke replied:
Well I'd really like to see this one fixed... Can't we just compare `.device()` of two tensors here? I just don't want to use a function that assigns threads to select the device.

apaszke commented on torch/csrc/autograd/input_metadata.h:
```diff
@@ -42,6 +52,13 @@ struct InputMetadata {
+  // but we have to fix all of the sites to adjust this way.  (Why is
+  // it this way?  In the old days we only had CUDA devices, and so
+  // we could just say -1 was CPU and everything else was CUDA.  This clearly
+  // doesn't hold anymore.)
```
This feels wrong. We should really use `Device` here, and the decision process that the engine uses to choose the thread shouldn't leak into here. Can't we just change those checks to use the real device, and contain the other logic? It doesn't seem to be _that_ hard. It's just pushing the call to `assign_tensor_to_autograd_thread` to a later point.

dlibenzi replied:
IMHO there should be to pinning of thread->device at thread boot time.
We should spawn an adequately sized pool, and let them fetch "work". The "work" will carry a  device, and the proper device guard should be set while that thread is processing that work.

ezyang replied:
@apaszke I attempted to do this months ago and there are a few tricky interaction points. I'd much rather do this refactor in a separate patch.

apaszke replied:
@dlibenzi it's actually not always beneficial to have thread pools there, hence this strategy.

@ezyang ok it doesn't have to be the full refactor but the comment above this one should be fixed before this lands. As is this patch is uber bug prone.

dlibenzi replied:
@apaszke For thread pool, I did not mean it in any dynamic way. Today PT is starting one thread per device, and binds it to a device. So for example if one have 16 devices and 4 cores, given that the work is CPU bound, we can have thrashing.
If PT instead starts min(num_devices, num_cores) threads, and does not bind the threads to devices at boot time, but instead let the device come in as part of the "work" the threads will pull from the queue, it seems the logic becomes simpler by removing the it-this-then-that logic in the thread init function, where PT creates the device guards.
The thread working loop can be like:
```
while (true) {
  auto work = thread_queue.get();
  auto device_guard = CreateDeviceGuard(work->device);
  process(work);
}
```
And the thread init function will lose a bunch of code.

apaszke reviewed (changes requested):
While it's going in a good direction I think this patch really confuses a few concepts. A proper cleanup shouldn't take too long, so let's do that before this lands.

ezyang commented:
@pytorchbot rebase this please

Commit:
Merge remote-tracking branch 'origin/master' into HEAD

ezyang committed:
Add some missing namespace qualifiers

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Merge branch 'xla_autograd' of github.com:dlibenzi/pytorch into xla_autograd

ezyang committed:
Redo engine modifications to generalize over all DeviceTypes

This fixes the MSNPU test, but now we require MSNPU to have
a DeviceGuardImpl (which makes sense.)

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Clarify comment.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
OK, I pushed an update to fix the MSNPU test, I generalized the device guard code along the way (so that we can handle ALL device types in the enum) and fixed up the comment.

dlibenzi commented:
@ezyang Can you ping me before landing this? This will break XLA, and we need to prep for it coming in, as we currently carry the original version of it:

https://github.com/pytorch/xla/tree/master/torch_patches

ezyang committed:
Test device update

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang commented:
@apaszke I'd prefer if you could take a pass on the new bits before I land :)

apaszke commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -203,22 +203,45 @@ Engine::Engine() = default;
-    if (at::hasHIP()) {
-      guard.reset_device(at::Device(at::DeviceType::HIP, device));
+    for (size_t i = 0; i < static_cast<size_t>(c10::DeviceType::COMPILE_TIME_MAX_DEVICE_TYPES); i++) {
+      auto* impl = c10::impl::device_guard_impl_registry[i].load();
```
Damn that's looking pretty deep into c10 internals

ezyang replied:
Yes, a pretty reasonable cleanup to ask for is to add some more functions to the C10 API and use those instead.

apaszke commented on torch/csrc/autograd/input_metadata.h:
```diff
@@ -42,6 +52,14 @@ struct InputMetadata {
 private:
   const at::Type* type_ = nullptr;
   at::DimVector shape_;
+  // This should be a 'Device'.  However, it is not for hysterical raisins,
```
😊

apaszke commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -217,7 +220,9 @@ const std::shared_ptr<Function>& Variable::grad_fn() const {
-      , diff_view_meta->base_.is_cuda() ? diff_view_meta->base_.get_device() : -1);
+      // TODO: stop throwing away useful device information here!
+      // Delay the autograd thread assignment to later.
+      , assign_tensor_to_autograd_thread(diff_view_meta->base_));
```
Can we call this function `drop_device_type` or `get_device_idx`?

apaszke reviewed (commented):
What's the value of COMPILE_TIME_MAX_DEVICE_TYPES?

I really don't like what's going on, but I guess we could land this, and I'll take a pass a cleaning up this mess later...

ezyang commented:
> `COMPILE_TIME_MAX_DEVICE_TYPES`

Currently it's 10.

apaszke commented:
I guess it would be nice to add a static assert that it never exceeds 16 or 32, so that we can avoid allocating huge stack arrays (and possibly overflowing the stack) in case someone increases it to incorporate an "infinite" number of devices.

ezyang committed:
Handle Adam's renaming request

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Merge remote-tracking branch 'origin/master' into xla_autograd

ezyang committed:
Actually check for Device.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Fix lint and more code cleanliness.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>
